yarn2nix: preserve top-level package.json resolutions field in workspace

purefn opened the pull request:
###### Motivation for this change
Fixes yarn package resolution errors as described in this issue https://github.com/nix-community/yarn2nix/issues/142#issuecomment-649049534.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

tcmal commented:
Currently having this same issue trying to build [youtube-music](https://github.com/th-ch/youtube-music). Is there any workaround / can this be merged?

purefn commented:
cc @SuperSandro2000

Lassulus commented:
hmm, this sadly has now a merge conflict

ambroisie commented:
I'm pretty sure i am also hitting this issue when trying to use `mkYarnModule` to build a Vue front-end, is there any workaround in the mean time?

Commit:
yarn2nix: preserve top-level package.json resolutions field in workspace

The `package.json` produced when building node_modules for a workspace
ignores the `resolutions` from the project `package.json`. This results
in dependencies being resolved in a way that conflicts with the
`yarn.lock` file. To fix this, we need to preserve the `resolutions`.

purefn commented:
Sorry for my absence. I've rebased on master and fixed the merge conflict.

It would be great if someone could review and merge this.

angerman commented:
Ran into this today. Can we get this merged?

Infinisil commented on pkgs/development/tools/yarn2nix-moretea/yarn2nix/default.nix:
```diff
@@ -170,6 +171,8 @@ in rec {
 
     packageGlobs = package.workspaces;
 
+    packageResolutions = package.resolutions;
```
Does this not cause trouble when there is no `resolutions` field?

purefn replied:
Hm good point. It would serialize to `null` in the resultant JSON, which yarn/npm may or may not handle correctly. I switched it to default to an empty attribute set.

Infinisil replied:
Not sure how serialization matters here, what gets serialized to `null`? If the attribute doesn't exist, Nix would've thrown an error.

purefn replied:
Sorry, yes, you're right. But it won't now.

infinisil reviewed (approved):
Looking good! Can you squash this into a single commit and change its name to the PR title? Notably `yarn2nix` should at least be mentioned in the commit message. Ideally the extended commit message would give some more details too

purefn commented:
@Infinisil squashed and revised the message. 

There is also a follow up issue that has come up recently. The way `yarn2nix` creates separate `node_modules` for each package in a workspace leads to some issues. Basically the way that a dependency is resolved for one package might be different than if all the packages were considered together. It seems `yarn2nix` should really be generating a single `node_modules` for a workspace that is used for each package.
